src/main/scala/RewriteView.scala

```diff
@@ -15,103 +18,82 @@ package fuselang
  * If `a` itself is a view, we keep rewriting it until we reach a true array.
  */
 object RewriteView {
-  import Syntax._
-  import CodeGenHelpers._
+  import Utils.State
 
-  // We can use a flat environment because the type checker makes sure things
-  // in the right scope.
   type T = Map[Id, List[Expr] => Expr]
 
-  def foldExprs(es: List[Expr])(implicit env: T): (List[Expr], T) =
-      es.foldLeft((List[Expr](), env))({
-        case ((idxsn, env), e) => {
-          val (e1, env1) = rewriteExpr(e)(env)
-          (e1 :: idxsn, env1)
-        }
-      })
-
-  def rewriteExpr(e: Expr)(implicit env: T): (Expr, T) = e match {
-    case EVar(_) | EInt(_, _) | EFloat(_) | EBool(_) | _:ERecAccess => e -> env
-    case eb@EBinop(_, e1, e2) => {
-      val (e1n, env1) = rewriteExpr(e1)
-      val (e2n, env2) = rewriteExpr(e2)(env1)
-      eb.copy(e1 = e1n, e2 = e2n) -> env2
-    }
-    case eaa@EArrAccess(arrId, idxs) => {
-      val (idxsn, env1) = foldExprs(idxs)
-      // If the array id is a view array rewrite it and recur
+  def rewriteExpr(e: Expr): State[T, Expr] = e match {
+    case EVar(_) | EInt(_, _) | EFloat(_) | EBool(_) | _:ERecAccess => State.unit(e)
+    case eb@EBinop(_, e1, e2) => for {
+      e1n <- rewriteExpr(e1)
+      e2n <- rewriteExpr(e2)
+    } yield eb.copy(e1 = e1n, e2 = e2n)
+    case eaa@EArrAccess(arrId, idxs) => State { env =>
```

Comment:
@sampsyo Is there a nice monadic function to represent what I'm doing here? This case requires access to the state (`env`) and the expression (`idxns`) to work.

Reply:
Hmm; I don't think so! I think this might be the simplest way to write this, AFAICT.